### PR TITLE
galaxy remove incorrect check for role path before extraction

### LINKED ISSUE #35217: ansible-galaxy install ignores -p switch when roles_path is configured
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

ansible-galaxy

##### ANSIBLE VERSION

```
ansible 2.5.0
  config file = /tmp/foo/ansible.cfg
  configured module search path = [u'<HOME>/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = <HOME>/.local/python2/ansible-devel/lib/python2.7/site-packages/ansible
  executable location = <HOME>/.local/python2/ansible-devel/bin/ansible
  python version = 2.7.14 (default, Sep 20 2017, 01:25:59) [GCC 7.2.0]
```

```
ansible 2.4.2.0
  config file = /tmp/foo/ansible.cfg
  configured module search path = [u'<HOME>/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = <HOME>/.local/python2/ansible-release/lib/python2.7/site-packages/ansible
  executable location = <HOME>/.local/python2/ansible-release/bin/ansible
  python version = 2.7.14 (default, Sep 20 2017, 01:25:59) [GCC 7.2.0]
```

##### CONFIGURATION

```
ANSIBLE_COW_SELECTION(/tmp/foo/ansible.cfg) = random
ANSIBLE_NOCOWS(env: ANSIBLE_NOCOWS) = True
ANSIBLE_PIPELINING(/tmp/foo/ansible.cfg) = True
ANSIBLE_SSH_ARGS(/tmp/foo/ansible.cfg) = -F ssh_config
DEFAULT_HOST_LIST(/tmp/foo/ansible.cfg) = [u'/tmp/foo/hosts']
DEFAULT_ROLES_PATH(/tmp/foo/ansible.cfg) = [u'/tmp/foo/roles', u'/tmp/roles', u'/tmp', u'/']
DEFAULT_VAULT_PASSWORD_FILE(/tmp/foo/ansible.cfg) = /tmp/foo/vault
HOST_KEY_CHECKING(/tmp/foo/ansible.cfg) = False
RETRY_FILES_ENABLED(/tmp/foo/ansible.cfg) = False
```

* **NOTE**: `roles_path` is configured to: `../:../../:../roles:roles/`

##### OS / ENVIRONMENT

N/A

##### SUMMARY

ansible-galaxy install ignores command line switch `-p` (`--roles-path`) if `roles_path` is configured via ansible.cfg. ansible-galaxy install keeps telling the role would be installed already.

##### STEPS TO REPRODUCE

```bash
install -d /tmp/ansible-galaxy-install-ignore-roles-path-bug
cd /tmp/ansible-galaxy-install-ignore-roles-path-bug
git clone https://github.com/retr0h/ansible-etcd.git
ansible-galaxy init roles-path-bug
cd roles-path-bug
cat <<EOF > ansible.cfg
[defaults]
roles_path = ../:../../:../roles:roles/
EOF
cat <<EOF > requirements.yml
---
- src: https://github.com/retr0h/ansible-etcd.git
  name: ansible-etcd
EOF
ansible-galaxy install -r requirements.yml -p roles/
```

```bash
ansible-galaxy install -r requirements.yml -p roles/
```

##### EXPECTED RESULTS

* Command line switches win over configuration values (**always!**)
* `roles/` directory getting created and role from requirements.yml installed inside `roles/` directory

##### ACTUAL RESULTS

* ansible-galaxy install states the role would already be installed and asks to use `--force` to replace it.

In my scenario, having roles and `"projects"` (bundling dependencies and inventories in git repositories) in the same directory, this might delete content on the disk which has not yet been pushed to a remote git service (as it happened to me yesterday).

```
- extracting ansible-etcd to /tmp/ansible-galaxy-install-ignore-roles-path-bug/ansible-etcd
 [WARNING]: - ansible-etcd was NOT installed successfully: the specified role
ansible-etcd appears to already exist. Use --force to replace it.

ERROR! - you can use --ignore-errors to skip failed roles and finish processing the list.
```

### PR DESCRIPTION
**Pull request by maxamillion**
##### SUMMARY

Fixes #35217

Currently lib/ansible/galaxy/role.py checks roles_path and will
ignore paths that don't currently exist which it should not because
the path will attempt to be created by the role extraction process
(which is valid and expected behavior) and if unable to write at the
time of role extraction into the role_path, that error will be
handled properly at that point in the process.

Signed-off-by: Adam Miller <admiller@redhat.com>

##### ISSUE TYPE

 - Bugfix Pull Request

##### COMPONENT NAME
lib/ansible/galaxy/role.py

##### ANSIBLE VERSION

```
ansible 2.5.0 (galaxy 083e4b812a) last updated 2018/01/23 17:34:22 (GMT -500)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/admiller/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/admiller/src/dev/ansible/lib/ansible
  executable location = /home/admiller/src/dev/ansible/bin/ansible
  python version = 2.7.14 (default, Jan 17 2018, 14:28:32) [GCC 7.2.1 20170915 (Red Hat 7.2.1-2)]

```

##### ADDITIONAL INFORMATION

Reproducer (from the Issue ticket):
```
install -d /tmp/ansible-galaxy-install-ignore-roles-path-bug
cd /tmp/ansible-galaxy-install-ignore-roles-path-bug
git clone https://github.com/retr0h/ansible-etcd.git
ansible-galaxy init roles-path-bug
cd roles-path-bug
cat <<EOF > ansible.cfg
[defaults]
roles_path = ../:../../:../roles:roles/
EOF
cat <<EOF > requirements.yml
---
- src: https://github.com/retr0h/ansible-etcd.git
  name: ansible-etcd
EOF
ansible-galaxy install -r requirements.yml -p roles/
```
Output Before Change:
```
- extracting ansible-etcd to /tmp/ansible-galaxy-install-ignore-roles-path-bug/ansible-etcd                                                    
 [WARNING]: - ansible-etcd was NOT installed successfully: the specified role ansible-etcd appears to already exist. Use --force to replace    
it.                                

ERROR! - you can use --ignore-errors to skip failed roles and finish processing the list. 
```

Output After Change:
```
- extracting ansible-etcd to /tmp/ansible-galaxy-install-ignore-roles-path-bug/roles-path-bug/roles/ansible-etcd
- ansible-etcd was installed successfully
```

### CODE REVIEW

**Comment by pilou-** [comment]
@maxamillion it seems a dedicated test would be useful: there are some tests related to `ansible-galaxy` in `test/units/cli/test_galaxy.py` and some cli tests in `test/integration/targets/ansible/runme.sh`.

**Comment by maxamillion** [comment]
@pilou- agreed, I'll add some test cases

File: lib/ansible/galaxy/role.py [hunk 1]
```diff
@@ -71,17 +71,11 @@ def __init__(self, galaxy, name, src=None, version=None, scm=None, path=None):
...
-            else:
-                # use the first path by default
-                self.path = os.path.join(galaxy.roles_paths[0], self.name)
-                # create list of possible paths
-                self.paths = [x for x in galaxy.roles_paths]
-                self.paths = [os.path.join(x, self.name) for x in self.paths]
+            # use the first path by default
+            self.path = os.path.join(galaxy.roles_paths[0], self.name)
+            # create list of possible paths
+            self.paths = [x for x in galaxy.roles_paths]
```
Comment (pilou-) [inline lib/ansible/galaxy/role.py:77]: In order to skip the first path which is used by default, should not that be `[x for x in galaxy.roles_paths[1:]]` ?

File: test/integration/targets/ansible/runme.sh [hunk 2]
```diff
@@ -15,3 +15,67 @@ ansible-config list | grep 'DEFAULT_REMOTE_USER'
...
+# BEGIN: Test case for ansible-galaxy
+#
+
+# Need a relative custom roles path for testing various scenarios of -p
+galaxy_relative_rolespath="./my/custom/roles/path"
+
+# Use geerlingguy's nginx galaxy role as the test dummy (it was the first
+# one that came to mind, this could be something else if someone prefers)
+galaxy_test_role="geerlingguy.nginx"
+galaxy_test_role_github="https://github.com/geerlingguy/ansible-role-nginx.git"
```
Comment (pilou-) [inline test/integration/targets/ansible/runme.sh]: In order to avoid network issues, the tarball could be pulled from [ansible-ci-files.s3.amazonaws.com](https://github.com/ansible/community/wiki/Testing%3A-progress-tracker#tests-that-use-external-dependencies-and-therefore-are-occasionally-unstable).

Comment (maxamillion) [reply test/integration/targets/ansible/runme.sh]: @pilou- how is pulling from one location on the internet less error prone than another?

File: lib/ansible/galaxy/role.py (hunk 1 above)
Comment (maxamillion) [reply lib/ansible/galaxy/role.py:77]: @pilou- the "recovery" logic for the `try`/`except` in the install method that handles the "first path found can not be written to, try the next one" depends on the path existing in `self.paths` because it checks the index and steps through the list. https://github.com/ansible/ansible/blob/42a0d7141339295cf7fd2f5ed05d7ec2ef5cb8c1/lib/ansible/galaxy/role.py#L337

File: test/integration/targets/ansible/runme.sh (hunk 2 above)
Comment (pilou-) [reply test/integration/targets/ansible/runme.sh]: AWS is used for some part of the CI. Some tests use the URL previously mentioned. It seems this is the [recommended location](https://github.com/ansible/community/wiki/Testing%3A-progress-tracker#tests-that-use-external-dependencies-and-therefore-are-occasionally-unstable).

Comment (maxamillion) [reply test/integration/targets/ansible/runme.sh]: @pilou- ah, learned something new today. Thanks :)

File: test/integration/targets/ansible/runme.sh [hunk 3]
```diff
@@ -15,3 +15,111 @@ ansible-config list | grep 'DEFAULT_REMOTE_USER'
 
 # 'view' command must fail when config file is missing
 ansible-config -c ./ansible-non-existent.cfg view && exit 1 || echo 'Failure is expected'
+
+################################################################################
+# BEGIN: Test case for ansible-galaxy
+#
+
+# Prep git, becuase it doesn't work inside a docker container without it
+git config --global user.email "tester@ansible.com"
```
Comment (mattclay) [inline test/integration/targets/ansible/runme.sh]: Using `--global` is dangerous, as it overwrites existing configuration. Avoid it if possible. If not, only set the values if they're not already set. Also make sure to add the `destructive` alias for the test target.

File: test/integration/targets/ansible/runme.sh [hunk 4]
```diff
@@ -15,3 +15,111 @@ ansible-config list | grep 'DEFAULT_REMOTE_USER'
 
 # 'view' command must fail when config file is missing
 ansible-config -c ./ansible-non-existent.cfg view && exit 1 || echo 'Failure is expected'
+
+################################################################################
```
Comment (mattclay) [inline test/integration/targets/ansible/runme.sh]: Please put the galaxy tests in their own test target named `ansible-galaxy`.

**Review by mattclay — CHANGES_REQUESTED** [review]

File: test/integration/targets/ansible/runme.sh (hunk 3 above)
Comment (maxamillion) [reply test/integration/targets/ansible/runme.sh]: @mattclay it runs in a docker container for the test, what's the potential fall out from `git config --global`?

Comment (mattclay) [reply test/integration/targets/ansible/runme.sh]: As discussed in IRC, the reason is that not everyone runs the tests in disposable docker containers. Tests that make changes outside the temporary directories of a test run should be as non-destructive as is practical and set the `destructive` alias to indicate that they do make such changes.

Another reason that I forgot to mention earlier is that it's always better if tests don't leave behind global changes when they finish, since those changes could interfere with a following test, even when using disposable containers. New containers are used per test run, not per test.

**Comment by maxamillion** [comment]
The OpenSUSE tests in shippable appear to be having network troubles pulling rpms from their repos, all other tests pass so I'm pretty sure it's unrelated to this change.

File: test/integration/targets/ansible/runme.sh [hunk 5]
```diff
@@ -15,3 +15,4 @@ ansible-config list | grep 'DEFAULT_REMOTE_USER'
 
 # 'view' command must fail when config file is missing
 ansible-config -c ./ansible-non-existent.cfg view && exit 1 || echo 'Failure is expected'
+
```
Comment (mattclay) [inline test/integration/targets/ansible/runme.sh]: Unnecessary change to existing test target.

File: test/integration/targets/ansible-galaxy/runme.sh [hunk 6]
```diff
@@ -0,0 +1,112 @@
+#!/usr/bin/env bash
+
+set -eux -o pipefail
+
+################################################################################
+# BEGIN: Test case for ansible-galaxy
+#
```
Comment (mattclay) [inline test/integration/targets/ansible-galaxy/runme.sh]: No need for start or end comments since this is the only test in the target.

File: test/integration/targets/ansible-galaxy/aliases [hunk 7]
```diff
@@ -0,0 +1 @@
+posix/ci/group3
```
Comment (mattclay) [inline test/integration/targets/ansible-galaxy/aliases]: Please include a trailing newline. It makes the diff cleaner in the future if another line is appended.

**Review by mattclay — CHANGES_REQUESTED** [review]

File: test/integration/targets/ansible-galaxy/aliases (hunk 7 above)
Comment (maxamillion) [reply test/integration/targets/ansible-galaxy/aliases]: @mattclay one of my vim plugins strips trailing whitespace, is this going to be a common request and I should alter my vim config or is this limited in scope to the `aliases`?

Comment (mattclay) [reply test/integration/targets/ansible-galaxy/aliases]: It's a general recommendation, not limited to `aliases` files.

Comment (maxamillion) [reply test/integration/targets/ansible-galaxy/aliases]: @mattclay I pushed a new commit with an empty line at the end of the file, but for some reason it's not showing in the diff on GitHub ... but when I open my local copy of the file it has an empty line at the bottom of the file and `git diff` says I'm good. I'm not sure what to do here.

Comment (mattclay) [reply test/integration/targets/ansible-galaxy/aliases]: I'm seeing two newlines now, but there should only be one.

**Review by mattclay — APPROVED** [review]
